security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignatureConfiguration.java

```diff
@@ -0,0 +1,267 @@
+     */
+    protected JWKSet jwkSetByUrl(String url) {
+        try {
+            return JWKSet.load(new URL(url));
+        } catch (FileNotFoundException e) {
+            if (LOG.isErrorEnabled()) {
+                LOG.error("FileNotFoundException loading JWK", e);
+            }
+        } catch (IOException e) {
+            if (LOG.isErrorEnabled()) {
+                LOG.error("IOException loading JWK", e);
+            }
+
+        } catch (ParseException e) {
+            if (LOG.isErrorEnabled()) {
+                LOG.error("ParseException loading JWK", e);
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Calculates a list of JWK matches for a JWT.
+     *
+     * Since the JWTSet is cached it attempts to refresh it (by calling its self recursive)
+     * if the {@param refreshKeysAttempts} is > 0.
+     *
+     * @param jwt A Signed JWT
+     * @param jwkSet A JSON Web Key Set
+     * @param refreshKeysAttempts Number of times to attempt refreshing the JWK Set
+     * @return a List of JSON Web Keys
+     */
+    protected List<JWK> matches(SignedJWT jwt, JWKSet jwkSet, int refreshKeysAttempts) {
+        if (jwkSet == null) {
+            return Collections.emptyList();
+        }
+        String keyId = jwt.getHeader().getKeyID();
+        if (keyId == null) {
+            return Collections.emptyList();
+        }
+        List<JWK> matches = new JWKSelector(
+                new JWKMatcher.Builder()
+                        .keyType(keyType)
+                        .keyID(keyId)
+                        .build()
+        ).select(jwkSet);
+
+        if (refreshKeysAttempts > 0 && matches.isEmpty()) {
+            this.jwkSet = jwkSetByUrl(url);
+            return matches(jwt, jwkSet, (refreshKeysAttempts - 1));
+        }
+        return matches;
+    }
+
+    /**
+     *
+     * @param jwk A JSON Web Key.
+     * @return a JWSVerifier for a JWK.
+     */
+    protected JWSVerifier verifiersForJwk(JWK jwk) {
+        if (jwk instanceof RSAKey) {
+            return jwsVerifierForRSAKey(((RSAKey) jwk));
+
+        } else if (jwk instanceof ECKey) {
+            return jwsVerifierForECKey((ECKey) jwk);
+        }
+        return null;
+    }
+
+    /**
+     *
+     * @param rsaKey A RSA Key.
+     * @return a JWSVerifier for a RSAKey.
+     */
+    protected JWSVerifier jwsVerifierForRSAKey(RSAKey rsaKey) {
+        RSAPublicKey rsaPublicKey = null;
+        try {
+            rsaPublicKey = rsaKey.toRSAPublicKey();
+        } catch (JOSEException e) {
+            if (LOG.isErrorEnabled()) {
+                LOG.error("JOSEException when retrieving RSA Public Key", e);
+            }
+        }
+        if (rsaPublicKey == null) {
+            return null;
+        }
+        return new RSASSAVerifier(rsaPublicKey);
+    }
+
+    /**
+     *
+     * @param ecKey A ECKey
+     * @return a JWSVerifier for a ECKey.
+     */
+    protected JWSVerifier jwsVerifierForECKey(ECKey ecKey) {
```

Comment:
Couldn't the 2 try catches be combined? The exception that gets logged should inform the user which step of the process failed